### PR TITLE
Make ListColumn.__init__ strict

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
This PR makes `ListColumn.__init__` strict putting restrictions on data, dtype, size and children so these columns cannot be constructed into to an invalid state. It also aligns the signature with the base class.

xref https://github.com/rapidsai/cudf/issues/16469

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/column/lists.py [hunk 1]
```diff
@@ -29,30 +29,46 @@
...
-        children=(),
+        data: None,
+        size: int,
+        dtype: ListDtype,
+        mask: Buffer | None = None,
+        offset: int = 0,
+        null_count: int | None = None,
+        children: tuple[NumericalColumn, ColumnBase] = (),  # type: ignore[assignment]
     ):
+        if data is not None:
```
Comment (bdice) [inline python/cudf/cudf/core/column/lists.py:54]: I see you want this to have the same signature as the base class. What does that actually get us? Does it help with dispatching somewhere?

More broadly, is there an issue tracking all these refactorings and explaining why they are being pursued?

File: python/cudf/cudf/core/column/lists.py [hunk 2]
```diff
@@ -29,30 +29,46 @@
...
+            raise ValueError("data must be None")
+        if not isinstance(dtype, ListDtype):
+            raise ValueError("dtype must be a cudf.ListDtype")
+        if not (
+            len(children) == 2
+            and isinstance(children[0], NumericalColumn)
+            # TODO: Enforce unsigned integers?
```
Comment (bdice) [inline python/cudf/cudf/core/column/lists.py:61]: libcudf uses `size_type` for offsets, which is defined as `int32_t` (it is signed).  https://github.com/rapidsai/cudf/blob/9d0c57a64d63d52182bd1c1e930180bf62404f1a/cpp/include/cudf/lists/lists_column_view.hpp#L122

File: python/cudf/cudf/core/column/lists.py (hunk 1 above)
Comment (mroeschke) [reply python/cudf/cudf/core/column/lists.py:54]: Sorry I added some context during today's meeting but you're right that I should probably be creating an issue for these themed PRs. I added some color in https://github.com/rapidsai/cudf/issues/16469

File: python/cudf/cudf/core/column/lists.py (hunk 2 above)
Comment (mroeschke) [reply python/cudf/cudf/core/column/lists.py:61]: Ah OK thanks!

**Review by bdice — APPROVED** [review]

**Pull request merged**